### ISSUE TITLE
Double Free / Segfault Issues

### ISSUE THREAD

**Issue by hartmir** [awslabs/aws-crt-cpp#181]
I found that cleanup inside gtest seems to have some issues. This code snippet demonstrates the double free issue - I have also had segfaults, but I have been unable to reproduce this with a minimal example.

```
TEST(JsonCleanupTest, testNoDoubleFree) {
  Aws::Crt::JsonObject obj;
  Aws::Crt::Vector<Aws::Crt::JsonObject> inner_array;
  for (int i = 0; i < 2000; i++) {
    Aws::Crt::JsonObject f;
    f.AsDouble((double)i);
    inner_array.push_back(f);
  }
  obj.WithArray("array", inner_array);
}
```

**Comment by rccarper** [comment]
Hello, sorry that you're running into that.  I'm not noticing a double free so far when running your snippet in our own testing framework.  Are you able to see what's getting double-freed?  Would it be possible for you to provide a call stack when you encounter the double-free?

**Comment by hartmir** [comment]
Sorry, I ran it in a clean project and also found it issue-free. I have been able to reproduce it by making it closer to my real use case. Please try the following:

```cpp
#include "gtest/gtest.h"

#include "aws/crt/JsonObject.h"
#include "aws/crt/Types.h"

struct DoubleFreeStruct {
  Aws::Crt::Vector<Aws::Crt::JsonObject> points;

  void AddPoint(float f) {
    Aws::Crt::JsonObject obj;
    obj.AsDouble(f);
    points.push_back(obj);
  }

  Aws::Crt::JsonObject ToJson() {
    Aws::Crt::JsonObject obj;
    obj.WithArray("points", points);
  }
};

TEST(DoubleFreeTest, testNoDoubleFree) {
  DoubleFreeStruct msg;
  for (int j = 0; j < 4000; j++) {
    msg.AddPoint(1.0);
  }
  auto json = msg.ToJson();
}

int main(int argc, char **argv) {
  ::testing::InitGoogleTest(&argc, argv);
  return RUN_ALL_TESTS();
}

```

**Comment by rccarper** [comment]
I do see a crash when running that in our test framework, but it's related to DoubleFreeStruct::ToJson not returning anything.  After making it return 'obj', it's running okay for me.  If you make that change on your side, are you still seeing a crash?

**Comment by hartmir** [comment]
Thanks for the response, @rccarper. I find the same thing. I tried to build an example closer to the real use case, but I cannot force the same error to occur. I did double check the original use case and it does return the object correctly, so that is not the error in the real use case. I suspect the difference is in the build environment, but I'd need to put more time in to track it down.

I think it is best to close the ticket for now, and I will reopen it once I have time to dig down to find the error.
